lib/events/dynamoevents/dynamoevents.go

```diff
@@ -1051,57 +1120,106 @@ func (l *Log) migrateDateAttribute(ctx context.Context) error {
 			if err != nil {
 				return trace.Wrap(err)
 			}
 
-			// Pull out the needed composite key attributes for the main table in a key object
-			// since this saves on bandwidth and Dynamo will reject anything with
-			// attributes other than those composing the primary key.
-			key := make(map[string]*dynamodb.AttributeValue)
-			key[keySessionID] = item[keySessionID]
-			key[keyEventIndex] = item[keyEventIndex]
+			item[keyDate] = dateAttribute
 
-			c := &dynamodb.UpdateItemInput{
-				TableName:                 aws.String(l.Tablename),
-				Key:                       key,
-				ExpressionAttributeValues: attributeMap,
-				UpdateExpression:          aws.String("SET CreatedAtDate = :date"),
+			wr := &dynamodb.WriteRequest{
+				PutRequest: &dynamodb.PutRequest{
+					Item: item,
+				},
 			}
 
-			_, err = l.svc.UpdateItem(c)
-			if err != nil {
-				return trace.Wrap(convertError(err))
+			writeRequests = append(writeRequests, wr)
+		}
+
+		for len(writeRequests) > 0 {
+			var top int
+			if len(writeRequests) > 25 {
+				top = 25
+			} else {
+				top = len(writeRequests)
 			}
 
-			if err := ctx.Err(); err != nil {
-				return trace.Wrap(err)
+			batch := writeRequests[:top]
+			writeRequests = writeRequests[top:]
+
+			// Don't exceed maximum workers.
+			for workerCounter.Load() >= maxMigrationWorkers {
+				select {
+				case <-time.After(time.Millisecond * 50):
+				case <-ctx.Done():
+					return trace.Wrap(ctx.Err())
+				}
 			}
+
+			workerCounter.Add(1)
+			go func() {
+				defer workerCounter.Sub(1)
+				amountProcessed := len(batch)
+
+				if err := l.uploadBatch(batch); err != nil {
+					workerErrors <- trace.Wrap(err)
+					return
+				}
+
+				total := totalProcessed.Add(int32(amountProcessed))
+				log.Infof("Migrated %d total events to 6.2 format...", total)
+			}()
 		}
 
 		// Setting the startKey to the last evaluated key of the previous scan so that
 		// the next scan doesn't return processed events.
 		startKey = scanOut.LastEvaluatedKey
 
-		total += *scanOut.Count
-		log.Infof("Migrated %d total events to 6.2 format...", total)
-
 		// If the `LastEvaluatedKey` field is not set we have finished scanning
 		// the entire dataset and we can now break out of the loop.
 		if scanOut.LastEvaluatedKey == nil {
 			break
 		}
 	}
 
+	// Wait for all workers to complete.
+	for workerCounter.Load() != 0 {
+		select {
+		case <-time.After(time.Millisecond * 50):
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		}
+	}
+
+	// Check for worker errors and escalate if found.
+	select {
+	case err := <-workerErrors:
+		return trace.Wrap(err)
+	default:
+	}
+
 	return nil
 }
 
+// uploadBatch creates or updates a batch of 25 events or less in one API call.
+func (l *Log) uploadBatch(writeRequests []*dynamodb.WriteRequest) error {
+	for {
+		c := &dynamodb.BatchWriteItemInput{
+			RequestItems: map[string][]*dynamodb.WriteRequest{l.Tablename: writeRequests},
+		}
+
+		out, err := l.svc.BatchWriteItem(c)
```

Comment:
`BatchWriteItem` indeed cannot perform conditional or partial updates in that sense. It can however batch replace items which is what we use it for here. We send the full data back up. I also made this assumption the first time I implemented this. The reason we need to batch here is to achieve sufficient speeds to migrate events. Any kind of PUT can also update an item via replacing it.